code/model/modelreplace.cpp

```diff
@@ -0,0 +1,409 @@
+	for (int i = 0; i < pm->n_models; i++) {
+		auto& submodel = pm->submodel[i];
+		for (auto& detail : submodel.details)
+			REPLACE_IF_EQ(detail, source, dest);
+		REPLACE_IF_EQ(submodel.first_child, source, dest);
+		REPLACE_IF_EQ(submodel.i_replace, source, dest);
+		for (auto& debris : submodel.live_debris)
+			REPLACE_IF_EQ(debris, source, dest);
+		REPLACE_IF_EQ(submodel.look_at_submodel, source, dest);
+		REPLACE_IF_EQ(submodel.my_replacement, source, dest);
+		REPLACE_IF_EQ(submodel.next_sibling, source, dest);
+		REPLACE_IF_EQ(submodel.parent, source, dest);
+	}
+}
+
+
+
+// Actual replacement operations
+
+VirtualPOFOperationAddSubmodel::VirtualPOFOperationAddSubmodel() {
+	required_string("+POF to Add:");
+	stuff_string(appendingPOF, F_FILESPEC);
+
+	required_string("+Source Subobject:");
+	stuff_string(subobjNameSrc, F_NAME);
+
+	if (optional_string("+Copy Children:")) {
+		stuff_boolean(&copyChildren);
+	}
+
+	if (optional_string("$Rename Subobjects:")) {
+		rename = make_unique<VirtualPOFOperationRenameSubobjects>();
+	}
+
+	required_string("+Destination Subobject:");
+	stuff_string(subobjNameDest, F_NAME);
+}
+
+void VirtualPOFOperationAddSubmodel::process(polymodel* pm, model_read_deferred_tasks& deferredTasks, model_parse_depth depth) const {
+	polymodel* appendingPM = new polymodel();
+	model_read_deferred_tasks appendingSubsys;
+	SCP_set<int> keepTextures;
+	read_model_file(appendingPM, appendingPOF.c_str(), 0, appendingSubsys, depth);
+
+	int src_subobj_no = -1;
+	for (int i = 0; i < appendingPM->n_models; i++) {
+		if (!stricmp(appendingPM->submodel[i].name, subobjNameSrc.c_str()))
+			src_subobj_no = i;
+	}
+
+	int dest_subobj_no = -1;
+	for (int i = 0; i < pm->n_models; i++) {
+		if (!stricmp(pm->submodel[i].name, subobjNameDest.c_str()))
+			dest_subobj_no = i;
+	}
+
+	if (src_subobj_no >= 0 && dest_subobj_no >= 0) {
+		create_family_tree(pm);
+		create_family_tree(appendingPM);
+
+		if (rename != nullptr)
+			rename->process(appendingPM, appendingSubsys, depth);
+
+		SCP_vector<int> to_copy_submodels;
+		bool has_name_collision = false;
+		if (copyChildren) {
+			model_iterate_submodel_tree(appendingPM, src_subobj_no, [&to_copy_submodels, &has_name_collision, pm, appendingPM](int submodel, int /*level*/, bool /*isLeaf*/) {
+				to_copy_submodels.emplace_back(submodel);
+				for (int i = 0; i < pm->n_models; i++) {
+					if (!stricmp(pm->submodel[i].name, appendingPM->submodel[submodel].name))
+						has_name_collision = true;
+				}
+				});
+		}
+		else {
+			to_copy_submodels.emplace_back(src_subobj_no);
+			for (int i = 0; i < pm->n_models; i++) {
+				if (!stricmp(pm->submodel[i].name, appendingPM->submodel[src_subobj_no].name))
+					has_name_collision = true;
+			}
+		}
+
+		if (!has_name_collision) {
+			//Make sure to keep old data
+			bsp_info* oldSubmodels = pm->submodel;
+			int old_n_submodel = pm->n_models;
+
+			//Realloc new submodel array of proper size
+			pm->n_models += (int)to_copy_submodels.size();
```

Comment:
I presume submodel post-processing, such as with `look_at_submodel_names` and `dock_parent_submodel_names`, will be done after all submodel copying is done?

It looks like the bit of code a few lines later with `change_submodel_numbers` handles this, except that it doesn't look like `parent_submodel` on the docking bays is handled.

Reply:
Yes, that is not yet handled.
For now, it doesn't matter, as only submodel data is copied. Since the original submodels keep their numbers, the docking bays on the old model stay valid. And since you cannot yet copy the docking bays of the other model, there is yet no need to do this.
But as the comment says: With Phase 2, _all_ submodel references in the polymodel will need to be processed, and not only those within the submodel structs

Reply:
Ok, sounds good.